### PR TITLE
Made variance `var<mat>` friendly (Issue #2101)

### PR DESCRIPTION
**Pull request by bbbales2**
## Summary

`variance` works with `var<mat>` 

## Release notes

`variance` modified to work with `var<mat>`

## Checklist

- [x] Math issue #2101

- [x] Copyright holder: Columbia University

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/rev/fun/variance.hpp
Comment (SteveBronder) [inline stan/math/rev/fun/variance.hpp:91]: [optional] One thing I've found to be a good bit faster, if we are doing what's essentially an accumulate and some other stuff, if you can do everything in a loop it's a good bit faster (much fewer reads/writes) Like here we could calculate the sum and the squared norm in a loop at once but tbh I'm not sure how used this function is so like we don't need to overoptimize everywhere

File: stan/math/rev/fun/variance.hpp
Comment (SteveBronder) [inline stan/math/rev/fun/variance.hpp:83]: [optional] Should we put these in `unlikely()` calls? idt they matter that much so no biggie

**Review by SteveBronder — DISMISSED** [review]
Cool! One yada yada if we want this a bit more optimized but I'm fine with it as is

File: stan/math/rev/fun/variance.hpp
Comment (SteveBronder) [inline stan/math/rev/fun/variance.hpp:73]: Docs for `T`!

File: stan/math/rev/fun/variance.hpp
Comment (bbbales2) [reply stan/math/rev/fun/variance.hpp:83]: I'll leave it to the branch predictor.

File: stan/math/rev/fun/variance.hpp
Comment (bbbales2) [reply stan/math/rev/fun/variance.hpp:91]: Oh we could compute `squaredNorm` and `x - mu` at the same time you're right.

Comment (bbbales2) [reply stan/math/rev/fun/variance.hpp:91]: I'll go back and change this since it's probably faster.

Comment (SteveBronder) [reply stan/math/rev/fun/variance.hpp:91]: idt we can do `x - mu` at the same time but we can do `squaredNorm` and `mu` at the same time

Comment (bbbales2) [reply stan/math/rev/fun/variance.hpp]: I think it works either way. THis way we subtract the mean before squaring. Maybe that keeps the numbers smaller? Not sure if it matters.

**Review by SteveBronder — DISMISSED** [review]
lgtm!

**Comment by bbbales2** [comment]
This pull never got merged and looked to have entered some sort of weird state, so I just merged in develop again. I'll merge it if it passes tests.

**Review by SteveBronder — APPROVED** [review]
Cool and good!

**Pull request merged**
